crypto/bn/bn_exp.c

```diff
@@ -1390,3 +1390,86 @@ int BN_mod_exp_simple(BIGNUM *r, const BIGNUM *a, const BIGNUM *p,
     bn_check_top(r);
     return ret;
 }
+
+/*
+ * This is a variant of modular exponentiation optimization that does
+ * parallel 2-primes exponentiation using 256-bit (AVX512VL) AVX512_IFMA ISA
+ * in 52-bit binary redundant representation.
+ * If such instructions are not available it falls back to two
+ * BN_mod_exp_mont_consttime() calls.
+ */
+int BN_mod_exp_mont_consttime_x2(BIGNUM *rr1, const BIGNUM *a1, const BIGNUM *p1,
+                                 const BIGNUM *m1, BN_MONT_CTX *in_mont1,
+                                 BIGNUM *rr2, const BIGNUM *a2, const BIGNUM *p2,
+                                 const BIGNUM *m2, BN_MONT_CTX *in_mont2,
+                                 BN_CTX *ctx)
+{
+    int ret = 0;
+
+#ifdef RSAZ_ENABLED
+    BN_MONT_CTX *mont1 = NULL;
+    BN_MONT_CTX *mont2 = NULL;
+
+    if (rsaz_avx512ifma_eligible() &&
```

Comment:
Since this is the only place that uses that function, I suggest you just look at the OPENSSL_ia32cap variable directly here instead of creating an assembler function for it.

Reply:
It is created on purpose. There is dependency on more or less modern assembler versions with IFMA instructions support to compile the code. And this function helps the code built with old assembler run correctly on IFMA capable CPU - perlasm script generates different versions of it depending on assembler version. 

Otherwise we could face the situation when the code compiled with old assembler and run on IceLake will fail, because OPENSSL_ia32cap based direct check will let this branch be executed.

Reply:
I see that rsaz-avx2.pl is doing the same.